qrenderdoc/Windows/ShaderViewer.cpp

```diff
@@ -2766,11 +2766,33 @@ RDTreeWidgetItem *ShaderViewer::makeSourceVariableNode(const SourceVariableMappi
         }
         else
         {
-          for(uint32_t a = 0; a < bind.arraySize; a++)
-            children.push_back(new RDTreeWidgetItem({
-                QFormatStr("%1[%2]").arg(localName).arg(a), QFormatStr("%1[%2]").arg(regNames).arg(a),
-                typeName, ToQStr(res.resources[a].resourceId),
-            }));
+          if(res.bindPoint.arraySize == bind.arraySize)
+          {
+            for(uint32_t a = 0; a < bind.arraySize; a++)
+              children.push_back(new RDTreeWidgetItem({
+                  QFormatStr("%1[%2]").arg(localName).arg(a),
+                  QFormatStr("%1[%2]").arg(regNames).arg(a), typeName,
+                  ToQStr(res.resources[a].resourceId),
+              }));
+          }
+          else
+          {
+            // The array was flattened when returned from the pipeline state. This is
+            // probably D3D12. Fetch the correct resource for each array index
```

Comment:
Can you remove the UI side changes here from this PR? The debugging part is fine and is unrelated (except for fixing issues from the same repro case), but I'd like to discuss this further before making any fixes here. I _really_ don't want to have API-specific hacks around the abstraction layer like this, since not only does it spread out anywhere that needs to access the resources but it also affects any other users of the abstraction layer like python scripts.

It might not be feasible to have a better fix due to D3D12 having cognitive dissonance about whether it supports resource arrays, but I'd like to try. Maybe at least we can try and contain this hack somewhere in the common pipeline state helpers.

Reply:
Sure, that's a good point. Will look into other ways to resolve this.